### ISSUE TITLE
tf.losses.softmax_cross_entropy is deviously inconsistent

### ISSUE THREAD

**Issue by ethereon** [tensorflow/tensorflow#6531]
There are at least three variants of `softmax_cross_entropy` in TensorFlow:

1. `tf.nn.softmax_cross_entropy_with_logits`

Accepts `logits` as the first argument, and `labels` as the second argument.

2. `tf.contrib.losses.softmax_cross_entropy`

Accepts `logits` as the first argument, and `onehot_labels` as the second argument. So far so good. Except this is deprecated, and displays the recommendation `Use tf.losses.softmax_cross_entropy instead`.

3. `tf.losses.softmax_cross_entropy`

Decides to switch things around for fun and have `onehot_labels` as the first argument and  `logits` as the second. Since `onehot_labels` and `logits` are identically shaped tensors, the call succeeds without any complaints (until something else fails as a consequence, like the gradient).

This inconsistency seems a bit error prone. If nothing else, perhaps the deprecation warning for  `tf.contrib.losses.softmax_cross_entropy` should include a heads up.

**Comment by michaelisard** [comment]
@dr4b does this need better doc?

**Comment by snnn** [comment]
Someone fell into this trap: https://github.com/tensorflow/models/pull/1053/

**Comment by allenlavoie** [comment]
Thank you for the report! Looks like the deprecation warning got updated by @alextp. I'm going to close this, since I don't see anything more we can do here given API compatibility.

**Comment by PatWie** [comment]
What's about 

```tf.losses.sparse_softmax_cross_entropy(labels,logits, ...)```
vs
```tf.nn.sparse_softmax_cross_entropy_with_logits(logits=logits, labels=label)```
?

Is there any reason to provide both functions? I understand that `tf.losses` also maintains `tf.GraphKeys.LOSSES`. But wouldn't it be more consistent (kargs vs. args, '_with_logits' vs. '') to only support a mixture by

```python
tf.metrics.sparse_softmax_cross_entropy_with_logits(labels=labels,logits=logits, 
                                                    ..., use_collection=False)
```

This way, `tf.nn.sparse_softmax_cross_entropy_with_logits` can be just wrapping `tf.metrics.sparse_softmax_cross_entropy_with_logits`.

It would be so much easier, if it would be possible to just write
```python
cost = tf.metrics.sparse_softmax_cross_entropy_with_logits(labels=labels,logits=logits)
prob = tf.nn.softmax(logits, name='prob')
pred = tf.arg_max(prob, 1, name='pred')
accuracy = tf.metrics.accuracy(labels=labels,pred=pred) # or even support logits
precision = tf.metrics.accuracy(labels=labels,pred=pred)
```
